src/net_processing.cpp

```diff
@@ -1736,24 +1744,167 @@ static void ProcessGetMWEBLeafset(CNode& pfrom, const CChainParams& chainparams,
+        }
+        return;
+    }
+
+    // Rewind leafset to block height
+    BlockValidationState state;
+    CCoinsViewCache temp_view(&chainman.ActiveChainstate().CoinsTip());
+    if (!ActivateArbitraryChain(state, temp_view, chainparams, pindex)) {
+        pfrom.fDisconnect = true;
+        return;
+    }
+
+    // Serve leafset to peer
+    MWEBLeafsetMsg leafset_msg(pindex->GetBlockHash(), temp_view.GetMWEBCacheView()->GetLeafSet()->ToBitSet());
+    connman.PushMessage(&pfrom, CNetMsgMaker(pfrom.GetCommonVersion()).Make(NetMsgType::MWEBLEAFSET, leafset_msg));
+}
+
+struct GetMWEBUTXOsMsg
+{
+    GetMWEBUTXOsMsg() = default;
+
+    SERIALIZE_METHODS(GetMWEBUTXOsMsg, obj)
+    {
+        READWRITE(obj.block_hash, VARINT(obj.start_index), obj.num_requested, obj.output_format);
+    }
+
+    uint256 block_hash;
+    uint64_t start_index;
+    uint16_t num_requested;
+    uint8_t output_format;
+};
+
+struct MWEBUTXOsMsg
+{
+    MWEBUTXOsMsg() = default;
+
+    SERIALIZE_METHODS(MWEBUTXOsMsg, obj)
+    {
+        READWRITE(obj.block_hash, VARINT(obj.start_index), obj.output_format, obj.utxos, obj.proof_hashes);
+    }
+
+    uint256 block_hash;
+    uint64_t start_index;
+    uint8_t output_format;
+    std::vector<NetUTXO> utxos;
+    std::vector<mw::Hash> proof_hashes;
+};
+
+static void ProcessGetMWEBUTXOs(CNode& pfrom, const ChainstateManager& chainman, const CChainParams& chainparams, CConnman& connman, const GetMWEBUTXOsMsg& get_utxos)
+{
+    if (get_utxos.num_requested > MAX_REQUESTED_MWEB_UTXOS) {
+        LogPrint(BCLog::NET, "getmwebutxos num_requested %u > %u, disconnect peer=%d\n", get_utxos.num_requested, MAX_REQUESTED_MWEB_UTXOS, pfrom.GetId());
+        if (!pfrom.HasPermission(PF_NOBAN)) {
+            pfrom.fDisconnect = true;
+        }
+        return;
+    }
+
+    static const std::set<uint8_t> supported_formats{
+        NetUTXO::HASH_ONLY,
+        NetUTXO::FULL_UTXO,
+        NetUTXO::COMPACT_UTXO};
+    if (supported_formats.count(get_utxos.output_format) == 0) {
+        LogPrint(BCLog::NET, "getmwebutxos output_format %u not supported, disconnect peer=%d\n", get_utxos.output_format, pfrom.GetId());
+        if (!pfrom.HasPermission(PF_NOBAN)) {
+            pfrom.fDisconnect = true;
+        }
+        return;
+    }
+
+    LOCK(cs_main);
+
+    if (chainman.ActiveChainstate().IsInitialBlockDownload()) {
+        LogPrint(BCLog::NET, "Ignoring getmwebutxos from peer=%d because node is in initial block download\n", pfrom.GetId());
+        return;
+    }
+
+    CBlockIndex* pindex = LookupBlockIndex(get_utxos.block_hash);
+    if (!pindex || !chainman.ActiveChain().Contains(pindex)) {
+        LogPrint(BCLog::NET, "Ignoring getmwebutxos from peer=%d because requested block hash is not in active chain\n", pfrom.GetId());
+        return;
+    }
+
+    // TODO: Add an outbound limit
+
+    // For performance reasons, we limit how many blocks can be undone in order to rebuild the leafset
```

Comment:
Some mentions of leafset from copy-pasting the other function, not sure whether the same rewind limit applies here. If so, the 'leafset' naming is confusing.